lib/puppet/type/file/source.rb

```diff
@@ -129,18 +129,6 @@ def checksum
       metadata && metadata.checksum
     end
 
-    # Look up (if necessary) and return local content.
-    def content
-      return @content if @content
-      raise Puppet::DevError, _("No source for content was stored with the metadata") unless metadata.source
-
-      tmp = Puppet::FileServing::Content.indirection.find(metadata.source, :environment => resource.catalog.environment_instance, :links => resource[:links])
```

Comment:
Need to pass `environment` and `links` when retrieving file content

Reply:
Oddly we used to pass `links` as an indirector request option when running `puppet apply` but we didn't send it as a query parameter when running `puppet agent` https://github.com/puppetlabs/puppet/blob/6.13.0/lib/puppet/type/file/source.rb#L295. For example, in 6.13.0, pluginsync sends `links=follow` in the `file_metadata` request, but not the `file_content` request:

```
<- "GET /puppet/v3/file_metadatas/plugins?environment=correct&links=follow&recurse=true&source_permissions=ignore&ignore=.svn&ignore=CVS&ignore=.git&ignore=.hg&checksum_type=md5 HTTP/1.1\r\nAccept: application/json, application/x-msgpack, text/pson\r\nX-Puppet-Version: 6.14.0\r\nAccept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3\r\nUser-Agent: Puppet/6.14.0 Ruby/2.3.8-p459 (x86_64-darwin18)\r\nHost: dear-playtime.delivery.puppetlabs.net:8140\r\n\r\n"
-> "HTTP/1.1 200 OK\r\n"
-> "Date: Thu, 20 Feb 2020 22:26:41 GMT\r\n"
-> "Content-Type: application/json;charset=utf-8\r\n"
-> "X-Puppet-Version: 6.14.0\r\n"
-> "Vary: Accept-Encoding, User-Agent\r\n"
-> "Content-Encoding: gzip\r\n"
-> "Content-Length: 277\r\n"
-> "\r\n"
...
Conn keep-alive
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/lib/puppet]/ensure: created
<- "GET /puppet/v3/file_content/plugins/puppet/foo.rb?environment=correct& HTTP/1.1\r\nAccept: application/octet-stream\r\nAccept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3\r\nUser-Agent: Ruby\r\nHost: dear-playtime.delivery.puppetlabs.net:8140\r\n\r\n"
````

Reply:
The `links` parameter is actually meaningless in the old code for two reasons. First, the `file_content` terminus ignores `links => manage` and always follows the link when retrieving content. Below notice that `find(..).content` returns the content of the file that the link points to:

```
(byebug) Puppet::FileServing::Content.indirection.find("puppet:///modules/filetest/link")
#<Puppet::FileServing::Content:0x00007fb1369a3780 @path="/Users/josh/.puppetlabs/etc/code/environments/production/modules/filetest/files/link", @links=:manage, @expiration=2020-02-21 14:10:17 -0800>
(byebug) Puppet::FileServing::Content.indirection.find("puppet:///modules/filetest/link").content
"link target content\n"
(byebug) Puppet::FileServing::Content.indirection.find("puppet:///modules/filetest/link", :links => :follow)
#<Puppet::FileServing::Content:0x00007fb136939718 @path="/Users/josh/.puppetlabs/etc/code/environments/production/modules/filetest/files/link", @links=:follow, @expiration=2020-02-21 14:10:24 -0800>
(byebug) Puppet::FileServing::Content.indirection.find("puppet:///modules/filetest/link", :links => :follow).content
"link target content\n"
```

Second, if `links => manage`, then puppet never tries to manage the "content" of the file. Instead it creates/updates the link in the file's ensure property https://github.com/puppetlabs/puppet/blob/6.13.0/lib/puppet/type/file/ensure.rb#L101-L104. So it's only if `links => follow` that we try to manage the file content via the source parameter.